File: 10.Task queues and background job/code/typescript/pattern_4_exactly_once_via_sqs_fifo_dedu_2.ts
// TypeScript, Idempotent multi-step task with full transaction rollback
import { PoolClient } from "pg";

interface DeleteAccountPayload {
    UserID: string;
}

interface Task {
    Payload(): string;
}

export class AccountHandler {
    private db: any; // e.g. Pool

    async handleDeleteAccount(t: Task): Promise<void> {
        const p: DeleteAccountPayload = JSON.parse(t.Payload());

        // Check if already deleted (idempotency guard)
        const exists = await this.db.userExists(p.UserID);
        if (!exists) {
            return; // already deleted on a previous attempt, ACK cleanly
        }

        // Wrap all DB writes in a single transaction
        return this.db.withTransaction(async (tx: PoolClient) => {
            const steps = [
                () => tx.query('DELETE FROM user_projects WHERE user_id = $1', [p.UserID]),
                () => tx.query('DELETE FROM user_sessions WHERE user_id = $1', [p.UserID]),
                () => tx.query('DELETE FROM user_assets WHERE user_id = $1', [p.UserID]),
                () => tx.query('DELETE FROM user_account WHERE id = $1', [p.UserID]),
            ];
            
            for (const step of steps) {
                await step(); // triggers full rollback on throw; task retried from scratch
            }
            // all steps succeeded -> commit -> ACK
        });
    }
}

File: 10.Task queues and background job/code/rust/pattern_4_exactly_once_via_sqs_fifo_dedu_2.rs
// Rust, Idempotent multi-step task with full transaction rollback
use sqlx::{PgPool, Postgres, Transaction};
use serde::Deserialize;

#[derive(Deserialize)]
struct DeleteAccountPayload {
    user_id: String,
}

pub struct Task {
    payload: String,
}

impl Task {
    fn payload(&self) -> &str {
        &self.payload
    }
}

pub struct AccountHandler {
    db: PgPool,
}

impl AccountHandler {
    pub async fn handle_delete_account(&self, t: &Task) -> Result<(), Box<dyn std::error::Error>> {
        let p: DeleteAccountPayload = serde_json::from_str(t.payload())?;

        // Check if already deleted (idempotency guard)
        let exists = self.user_exists(&p.user_id).await?;
        if !exists {
            return Ok(()); // already deleted on a previous attempt, ACK cleanly
        }

        // Wrap all DB writes in a single transaction
        let mut tx = self.db.begin().await?;
        
        self.delete_user_projects(&mut tx, &p.user_id).await?;
        self.delete_user_sessions(&mut tx, &p.user_id).await?;
        self.delete_user_assets(&mut tx, &p.user_id).await?;
        self.delete_user_account(&mut tx, &p.user_id).await?;
        
        tx.commit().await?; // all steps succeeded -> commit -> ACK
        
        Ok(())
    }

    async fn user_exists(&self, _user_id: &str) -> Result<bool, sqlx::Error> {
        Ok(true) // placeholder
    }

    async fn delete_user_projects(&self, _tx: &mut Transaction<'_, Postgres>, _user_id: &str) -> Result<(), sqlx::Error> { Ok(()) }
    async fn delete_user_sessions(&self, _tx: &mut Transaction<'_, Postgres>, _user_id: &str) -> Result<(), sqlx::Error> { Ok(()) }
    async fn delete_user_assets(&self, _tx: &mut Transaction<'_, Postgres>, _user_id: &str) -> Result<(), sqlx::Error> { Ok(()) }
    async fn delete_user_account(&self, _tx: &mut Transaction<'_, Postgres>, _user_id: &str) -> Result<(), sqlx::Error> { Ok(()) }
}
